doc/designs/functions-for-explicitly-fetched-signature-algorithms.md

```diff
@@ -0,0 +1,155 @@
+Functions for explicitly fetched PKEY algorithms
+================================================
+
+Quick background
+----------------
+
+There are several proposed designs that end up revolving around the same
+basic need, explicitly fetched signature algorithms.  The following method
+type is affected by this document:
+
+- `EVP_SIGNATURE`
+
+Public API - Add variants of `EVP_PKEY_CTX` functionality
+---------------------------------------------------------
+
+Through OTC discussions, it's been determined that the most suitable APIs to
+touch is the collection of `EVP_PKEY_` functions.  In this case, it involves
+`EVP_PKEY_sign()`, `EVP_PKEY_verify()`, `EVP_PKEY_verify_recover()` with
+associated functions.  They can be extended to accept an explicitly fetched
+algorithm of the right type, and to be able to process infinite amounts of
+data if the fetched algorithm permits it (typically, algorithms like ED25519
+or RSA-SHA256).
```

Comment:
As @bernd-edlinger and @davidben wrote in this https://github.com/openssl/openssl/pull/23240#issuecomment-1916968985 and following comments this is potentially confusing to users and we might want to separate the API calls for _signing a message digest_ from calls for _signing a message_.

On the other hand the EVP_DigestSign set of API calls are misdesigned and ugly so my proposal is to:

1. Add an alias EVP_PKEY_sign_digest() to the existing EVP_PKEY_sign() call, and EVP_PKEY_verify_digest() to the existing EVP_PKEY_verify() call.
2. Add functions EVP_PKEY_sign_message(), EVP_PKEY_verify_message() that would be implemented by these ED25519, RSA-SHA256 or similar algorithms.

Reply:
Depending on how you read it, *message* is quite confusing too.  If you read PKCS#1 carefully, you'll note that RSASP1 and RSAVP1 also take a (limited size) message, *represented* as an integer 0 → n - 1 (they talk about "message representative", which I read to mean that the integer is a representation of the message, a limited size octet string that was passed to O2ISP, or that will get passed through I2OSP).

Reply:
Thinking a bit about it, it could be that *primitive* and *data* provide a better distinction, with the semantics that *primitive* only allow a one-shot style of operation, while *data* allows a stream style of operation.  So we'd end up with:

- `EVP_PKEY_sign_primitive_init()` and `EVP_PKEY_sign_primitive()` (with the old `EVP_PKEY_sign_init()` and `EVP_PKEY_sign()` as aliases)
- `EVP_PKEY_sign_data_init()`, `EVP_PKEY_sign_data_update()` and `EVP_PKEY_sign_data_final()`

... and similar for any other stuff we want to give the same treatment.

Reply:
Yeah, that works for me.

Reply:
Now we need to think about the provider side interface.

Reply:
Could we just map the new APIs to the digest_sign_init()/digest_verify_init() provider functions with adding appropriate aliases through macros and add just new calls as needed?

Reply:
I've changed things around a bit, hopefully legibly